nucypher/network/protocols.py

```diff
@@ -89,3 +93,140 @@ def __radd__(self, other):
+            retention: self.severe_warning
+        }
+
+        self._ursula = ursula
+        self._sample_size = sample_size
+        self._records = deque(maxlen=retention)
+
+        self.__start_time = None
+        self.__task = LoopingCall(self.maintain)
+
+    def mild_warning(self) -> None:
+        self.log.info(f'{self._ursula.rest_url} is unreachable')
+
+    def medium_warning(self) -> None:
+        self.log.warn(f'{self._ursula.rest_url} is unreachable')
+
+    def severe_warning(self) -> None:
+        self.log.warn(f'{self._ursula.rest_url} is unreachable')
+        try:
+            raise self.Unreachable(f'{self._ursula} is unreachable.')
+        finally:
+            if reactor.running:
+                reactor.stop()
+
+    def handle_measurement_errors(self, *args, **kwargs) -> None:
+        failure = args[0]
+        cleaned_traceback = failure.getTraceback().replace('{', '').replace('}', '')  # FIXME: Amazing.
+        self.log.warn("Unhandled error during availability check: {}".format(cleaned_traceback))
+        failure.raiseException()
+
+    @property
+    def status(self) -> bool:
+        """Returns current indication of availability"""
+        if not self._records:
+            return
+        return self.score > self.sensor_sensitivity
+
+    @property
+    def running(self) -> bool:
+        return self.__task.running
+
+    def start(self, now: bool = False):
+        if not self.running:
+            self.__start_time = maya.now().epoch
+            d = self.__task.start(interval=self.interval, now=now)
+            d.addErrback(self.handle_measurement_errors)
+
+    def stop(self) -> None:
+        if self.running:
+            self.__task.stop()
+
+    def maintain(self) -> None:
+        self.log.debug(f"Starting new sensor maintenance round")
+        if not self._ursula.known_nodes:
+            # If there are no known nodes, skip this round...
+            # ... but not for longer than the maximum allotted alone time
+            if not self._ursula.lonely:
+                now = maya.now().epoch
+                delta = now - self.__start_time
+                if delta >= self.MAXIMUM_ALONE_TIME:
+                    self.severe_warning()
+            return
+        if len(self._ursula.known_nodes) < self.sample_size:
+            # If there are fewer known nodes then the sample size, skip this round
+            # TODO: How long can this be allowed for?
+            return
+        result = self.measure()
+        self.record(result)
+        self.issue_warnings()
+
+    @property
+    def score(self) -> float:
+        failed_records = list(record.result for record in self._records if not record.result)
+        return len(failed_records) / self.retention
+
+    def issue_warnings(self) -> None:
+        for threshold, action in self.warnings.items():
+            if self.score >= threshold:
+                action()
+
+    def sample(self, quantity: int) -> list:
```

Comment:
Fine for now but... Is it possible to use Alice's sample?

Though it requires some thinking, so let's start with this, simple one

Reply:
Yes, it's possible - What do you think are the advantages?

Reply:
Just a more objective measurement. But I think, the advantages are diminishingly small, as I'm thinking now

Reply:
They have a minimum stake presumably anyway, that's good enough

Reply:
Do they have the minimum stake though? Or could they be stakeless nodes?
If they can be stakeless, someone can spin up a bunch of nodes w/o NU and shut down the whole network by always responding that every node is down

Reply:
We're depending on the LL to verify staking here.